### PR TITLE
add date-based versioning

### PR DESCRIPTION
**Pull request by retokromer**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?

As suggested in https://github.com/Homebrew/brew/issues/2439 adds date-based versioning (RFC 3339).

### CODE REVIEW

File: Library/Homebrew/version.rb [hunk 1]
```diff
@@ -408,6 +408,11 @@ def self._parse(spec)
     # e.g. http://www.ijg.org/files/jpegsrc.v8d.tar.gz
     m = /\.v(\d+[a-z]?)/.match(stem)
     return m.captures.first unless m.nil?
+
+    # date-based versioning
+    # e.g. ltopers-v2017-04-14.tar.gz
```
Comment (MikeMcQuaid) [inline Library/Homebrew/version.rb]: Please add a test for this case.

Comment (MikeMcQuaid) [inline Library/Homebrew/version.rb]: (in https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/version_spec.rb)

Comment (retokromer) [inline Library/Homebrew/version.rb]: I’ve to delve a little deeper into that, but cannot do this immediately.

Comment (MikeMcQuaid) [inline Library/Homebrew/version.rb]: It's because the ordering means this will be caught by another regex first.

Comment (retokromer) [inline Library/Homebrew/version.rb]: Thank you for the hint!

File: Library/Homebrew/version.rb [hunk 2]
```diff
@@ -344,6 +344,11 @@ def self._parse(spec)
     m = /[-v]((?:\d+\.)*\d+)$/.match(spec_s)
     return m.captures.first unless m.nil?
 
+    # date-based versioning
+    # e.g. ltopers-v2017-04-14.tar.gz
+    m = /-v?((?:\d{4}-\d{2}-\d{2}))/.match(stem)
```
Comment (reitermarkus) [inline Library/Homebrew/version.rb]: You can remove the inner uncaptured group.

File: Library/Homebrew/version.rb [hunk 3]
```diff
@@ -344,6 +344,11 @@ def self._parse(spec)
     m = /[-v]((?:\d+\.)*\d+)$/.match(spec_s)
     return m.captures.first unless m.nil?
 
+    # date-based versioning
+    # e.g. ltopers-v2017-04-14.tar.gz
+    m = /-v?(?:\d{4}-\d{2}-\d{2})/.match(stem)
```
Comment (reitermarkus) [inline Library/Homebrew/version.rb]: Now you removed the outer group, it should be `/-v?(\d{4}-\d{2}-\d{2})/`

Comment (retokromer) [inline Library/Homebrew/version.rb]: Sorry!

**Comment by MikeMcQuaid** [comment]
Looks good, thanks @retokromer!

**Comment by ilovezfs** [comment]
Seems like this cannot handle `re2`'s date-based versions (https://github.com/google/re2/releases). For instance `2017-05-01` parses as `05-01`.

**Comment by retokromer** [comment]
Hmm… That was the behavior before the change! Possibly the regex needs to be moved before https://github.com/retokromer/brew/blob/3585dfd9bd84005ba20e4a97802dea13fb82c63e/Library/Homebrew/version.rb#L335 or so. Sorry, I’m travelling today, but will delve into this later.

**Comment by ilovezfs** [comment]
@retokromer cool, thanks. Also, I notice that
```
irb(main):001:0> Version.new("2017-05-01") > Version.new("20170401`")
=> false
```
It might be useful to have that compare as true to avoid having to bump `version_scheme` every time we switch from a `20170401` style version to a `2017-05-01` style version.

**Comment by retokromer** [comment]
@ilovezfs Moving the test at the beginning of the tests seems to work as expected. Now I’ll try to find and understand the comparison code.

**Comment by MikeMcQuaid** [comment]
Problem with moving it to the beginning is the potential to break versions on other formulae unless the regex is very specific.

**Comment by retokromer** [comment]
In my opinion, it is indeed very specific: `-v?(\d{4}-\d{2}-\d{2})`, i.e. the hard coded date as in RFC 3339.
The problem is that the current first regex matches the second dash of the date…
I’m still travelling, but I’ll try to propose a solution that solves also the remark above https://github.com/Homebrew/brew/pull/2490#issuecomment-298304347.

**Comment by retokromer** [comment]
For the moment, I made the PR https://github.com/Homebrew/brew/pull/2591 with just moving this regex at the beginning, as I don’t see other regexes with two dashes that could conflict.

**Pull request merged**